### PR TITLE
25 expand test poisson to 2d and other boundary conditions

### PR DESCRIPTION
**Pull request by gusespinola**
Current test_poisson.m contains

- [x] `test_poisson_1d_dir_bc`: 1d Poisson equation with Dirichlet bc at left & right nodes.
- [x] `test_poisson_1d_dir_neu_bc`: 1d Poisson equation with Dirichlet bc at the left node and a Neumann bc at the right node.
- [x] `test_poisson_1d_robin_bc`: 1d Poisson equation with Robin bc at both sides.
- [x] `test_poisson_2d_dir_bc`: 2d Poisson equation with Dirichlet bc at all sides.

Work in progress:
- [x] `test_poisson_2d_left_to_right_flow`: 2d Poisson equation with Dirichlet boundaries at left `(u = 1)` and right `(u = 0)` and no-flow at the top and bottom. Note that `f(x) = 0`. The exact solution is linear on `u`.

### CODE REVIEW

File: tests/test_poisson.m [hunk 1]
```diff
@@ -83,3 +88,301 @@ function test_poisson_one_dimension_dirichlet_bc()
...
+end
+
+function test_poisson_2d_dir_bc()
+    % Test solvers for a 2D-FDM linear system with 
+    % Dirichlet boundaries.
+    %
+    % Description:
+    % ============
+    %
+    % Solve a 2D-Poisson equation at [0, 1] x [0, 1]
```
Comment (jhabriel) [inline tests/test_poisson.m:233]:
```suggestion
    % Solve a 2D-Poisson equation in [0, 1] x [0, 1]
```

File: tests/test_poisson.m [hunk 2]
```diff
@@ -83,3 +88,301 @@ function test_poisson_one_dimension_dirichlet_bc()
...
+    h = ( aEnd - aStart ) / (NODES1D - 1); %;
+
+    % Left-hand side (LHS) matrix 'A'
+    % LHS is composed by the finite-difference coefficients for u''(x)
+    M = 4*eye(INNODES1D) - diag(ones(INNODES1D-1,1), 1) ...
+        -diag(ones(INNODES1D-1, 1), -1);
+    N = -1*eye(INNODES1D);
+    Z = zeros(INNODES1D);
+    A = [M N Z; N M N; Z N M];
+    % For INNODES1D > 3, previous line requires an external 'blktridiag' script
```
Comment (jhabriel) [inline tests/test_poisson.m:264]:
```suggestion
    % For INNODES1D > 3, previous line requires the built-in 'blktridiag' function
```

File: tests/test_poisson.m [hunk 3]
```diff
@@ -83,3 +88,301 @@ function test_poisson_one_dimension_dirichlet_bc()
...
+    tol = 1e-9;
+    maxit = 100;
+    mInitial = 3;
+    uPD_GMRES = pd_gmres(A, b, mInitial, [], [], tol, maxit);
+    
+    % Assert whether the pd_gmres solution match the exact knonw solution
+    assertElementsAlmostEqual(uPD_GMRES, uExact)
+
+end
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Comment (jhabriel) [inline tests/test_poisson.m:302]: Please, remove this line and the one below.

File: tests/test_poisson.m [hunk 4]
```diff
@@ -83,3 +88,301 @@ function test_poisson_one_dimension_dirichlet_bc()
...
+% NEW! (2024)
+function test_poisson_2d_dir_neu_bc()
+    % Alternative name:
+    % =================
+    % function test_poisson_left_to_right_flow()
```
Comment (jhabriel) [inline tests/test_poisson.m:307]: I would suggest the name: `test_poisson_2d_left_to_right_flow()`

File: tests/test_poisson.m [hunk 5]
```diff
@@ -83,3 +88,301 @@ function test_poisson_one_dimension_dirichlet_bc()
...
+    %
+    % Description:
+    % ============
+    % Test solvers for a 2D-FDM linear system with 
+    % Dirichlet-Neumann boundaries.
+    %
+    % 2-D Poisson equation with Dirichlet boundaries at left (u = 1) 
+    % and right (u = 0) and no-flow at the top and bottom. 
+    % Note that f(x) = 0. The exact solution is linear on u.
```
Comment (jhabriel) [inline tests/test_poisson.m:316]:
```suggestion
    % Note that f(x, y) = 0. The exact solution u is linear.
```

File: tests/test_poisson.m [hunk 6]
```diff
@@ -83,3 +88,301 @@ function test_poisson_one_dimension_dirichlet_bc()
...
+    %
+    % Description:
+    % ============
+    %
+    % Solve a 2D-Poisson equation at [0, 1] x [0, 1]
```
Comment (jhabriel) [inline tests/test_poisson.m:321]:
```suggestion
    % Solve a 2D-Poisson equation in [0, 1] x [0, 1]
```

**Review by jhabriel — APPROVED** [review]
Fantastic job @gusespinola. 

I will approve this PR, but please take care of two things before merging.

1. The file `test_report.log` is still present as a new file.
2. Please remove commented lines 267, 268, 354, 355, 356 from `test_poisson.m`

**Comment by gusespinola** [comment]
Sorry, I hadn't noticed that. Done!
